lib/common/asar.js

```diff
@@ -39,125 +35,93 @@
   // Separate asar package's path from full path.
   const splitPath = function (p) {
     // shortcut to disable asar.
-    if (isAsarDisabled()) {
-      return [false]
-    }
-
-    if (Buffer.isBuffer(p)) {
-      p = p.toString()
-    }
-
-    if (typeof p !== 'string') {
-      return [false]
-    }
-
-    if (p.substr(-5) === '.asar') {
-      return [true, p, '']
-    }
+    if (isAsarDisabled()) return [false]
+    if (Buffer.isBuffer(p)) p = p.toString()
+    if (typeof p !== 'string') return [false]
+    if (p.substr(-5) === '.asar') return [true, p, '']
 
     p = path.normalize(p)
+
     const index = p.lastIndexOf('.asar' + path.sep)
-    if (index === -1) {
-      return [false]
-    }
+    if (index === -1) return [false]
+
     return [true, p.substr(0, index + 5), p.substr(index + 6)]
   }
 
   // Convert asar archive's Stats object to fs's Stats object.
   let nextInode = 0
 
   const uid = process.getuid != null ? process.getuid() : 0
-
   const gid = process.getgid != null ? process.getgid() : 0
 
   const fakeTime = new Date()
+  const msec = (date) => (date || fakeTime).getTime()
 
   const asarStatsToFsStats = function (stats) {
-    return {
-      dev: 1,
-      ino: ++nextInode,
-      mode: 33188,
-      nlink: 1,
-      uid: uid,
-      gid: gid,
-      rdev: 0,
-      atime: stats.atime || fakeTime,
-      birthtime: stats.birthtime || fakeTime,
-      mtime: stats.mtime || fakeTime,
-      ctime: stats.ctime || fakeTime,
-      size: stats.size,
-      isFile: function () {
-        return stats.isFile
-      },
-      isDirectory: function () {
-        return stats.isDirectory
-      },
-      isSymbolicLink: function () {
-        return stats.isLink
-      },
-      isBlockDevice: function () {
-        return false
-      },
-      isCharacterDevice: function () {
-        return false
-      },
-      isFIFO: function () {
-        return false
-      },
-      isSocket: function () {
-        return false
-      }
-    }
+    const { Stats, constants } = require('fs')
+
+    let mode = constants.S_IROTH ^ constants.S_IRGRP ^ constants.S_IRUSR ^ constants.S_IWUSR
```

Comment:
To explain to reviewers, this `mode` value generated is the same as `33188` (the old value) but uses the correct file type mask.  Previously the file type was always set to `S_IFREG`, if we use the node `Stats` class here we can just set the correct flag in the mask and the node implementation will do the work for us